crates/hotshot-state-prover/src/circuit.rs

```diff
@@ -116,236 +109,244 @@ impl AsRef<[Variable]> for LightClientStateVar {
     }
 }
 
-#[derive(Clone, Debug)]
-pub struct StateUpdateBuilder<F: RescueParameter>(PhantomData<F>);
-
-impl<F> StateUpdateBuilder<F>
+/// A function that takes as input:
+/// - a list of stake table entries (`Vec<(SchnorrVerKey, Amount)>`)
+/// - a bit vector indicates the signers
+/// - a list of schnorr signatures of the updated states (`Vec<SchnorrSignature>`), default if the node doesn't sign the state
+/// - updated light client state (`(view_number, block_height, block_comm, fee_ledger_comm, stake_table_comm)`)
+/// - a quorum threshold
+/// Lengths of input vectors should not exceed the `STAKE_TABLE_CAPACITY`.
+/// The list of stake table entries, bit indicators and signatures will be padded to the `STAKE_TABLE_CAPACITY`.
+/// It checks that
+/// - the signer's accumulated weight exceeds the quorum threshold
+/// - the stake table corresponds to the one committed in the light client state
+/// - all signed schnorr signatures are valid
+/// and returns
+/// - A circuit for proof generation
+/// - A list of public inputs for verification
+/// - A PlonkError if any error happens when building the circuit
+pub(crate) fn build_state_verifier_circuit<F, P, STIter, BitIter, SigIter>(
+    stake_table_entries: STIter,
+    signer_bit_vec: BitIter,
+    signatures: SigIter,
+    lightclient_state: &LightClientState<F>,
+    threshold: &U256,
+) -> Result<(PlonkCircuit<F>, Vec<F>), PlonkError>
 where
     F: RescueParameter,
+    P: TECurveConfig<BaseField = F>,
+    STIter: IntoIterator,
+    STIter::Item: Borrow<(SchnorrVerKey<P>, U256)>,
+    STIter::IntoIter: ExactSizeIterator,
+    BitIter: IntoIterator,
+    BitIter::Item: Borrow<bool>,
+    BitIter::IntoIter: ExactSizeIterator,
+    SigIter: IntoIterator,
+    SigIter::Item: Borrow<Signature<P>>,
+    SigIter::IntoIter: ExactSizeIterator,
```

Comment:
Can I also propose to change `STIter` to `ST::IntoIter where ST: StakeTableScheme`? 
I want to avoid the extra iterating over stake table entries just to build an input for this function from the outside:

```rust
// I'd prefer avoid building `stake_table_entries`
    let stake_table_entries = stake_table
        .try_iter(SnapshotVersion::LastEpochStart)
        .unwrap()
        .map(|(_, stake_amount, schnorr_key)| (schnorr_key, stake_amount))
        .collect::<Vec<_>>();
    let (circuit, public_inputs) = build_state_verifier_circuit(
        stake_table_entries, // here I want to directly pass in `st.try_iter()?`

``` 

Is this possible?

Reply:
Borrow is for more generic input types, e.g. iterator of references.
I add an extra bound `ST::IntoIterator: ExactIterator`, so I can directly pass in the `map` without collecting the entries. 
check febaee5

Reply:
Right, but we do we need the flexibility of accepting iterator of references for now? 

Let's think about the actual usage, the prover will have these lay flat out in its memory, everything owned, 
to even have a vector of reference the code will even need to build a separate vector, which I don't think the prover code would do.

the `.iter()` by default only takes references of `Item` and won't consume the ownership to start with. So I prefer removing these heavy trait bounds including the `Borrow` here.

Reply:
First, there's no harm in make it more generic right now. 
Second, in the test where you are building multiple good or bad circuits, if you take the ownership, you need then construct multiple stake table entry lists. Doesn't look pretty that way